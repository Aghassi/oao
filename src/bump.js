// @flow

import { setIn } from 'timm';
import { mainStory } from 'storyboard';
import { readAllSpecs } from './utils/readSpecs';
import { DEP_TYPES } from './utils/constants';
import { parseDep } from './utils/helpers';
import writeSpecs from './utils/writeSpecs';
import { exec } from './utils/shell';

type Options = {
  src: string,
  ignoreSrc?: string,
  link: ?string,
};

const run = async (deps: Array<string>, opts: Options) => {
  const { src, ignoreSrc, link: linkPattern } = opts;
  const allSpecs = await readAllSpecs(src, ignoreSrc);
  const pkgNames = Object.keys(allSpecs);

  // Determine correct version for each dep
  const versions = {};
  for (let i = 0; i < deps.length; i++) {
    const dep = deps[i];
    const { name, version } = parseDep(dep);
    let finalVersion = version;
    if (!finalVersion) {
      if (pkgNames.indexOf(dep) >= 0) {
        finalVersion = `^${allSpecs[dep].specs.version}`;
      } else if (linkPattern && new RegExp(linkPattern).test(dep)) {
        finalVersion = '*';
      } else {
        const { stdout } = await exec(`npm info ${dep} version`, {
          logLevel: 'trace',
        });
        finalVersion = `^${stdout.trim()}`;
      }
    }
    versions[name] = finalVersion;
  }
  mainStory.info('New versions of these packages:', { attach: versions });

  // Update all package.json files with this version
  pkgNames.forEach(pkgName => {
    const { specPath, specs: prevSpecs } = allSpecs[pkgName];
    deps.forEach(depName => {
      let nextSpecs = prevSpecs;
      DEP_TYPES.forEach(type => {
        const depsOfType = nextSpecs[type] || {};
        if (depsOfType[depName] != null) {
          nextSpecs = setIn(nextSpecs, [type, depName], versions[depName]);
        }
      });
      if (nextSpecs !== prevSpecs) writeSpecs(specPath, nextSpecs);
    });
  });
};

export default run;
