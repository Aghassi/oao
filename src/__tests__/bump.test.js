/* eslint-env jest */
/* eslint-disable global-require, import/no-dynamic-require */

import bump from '../bump';

jest.mock('../utils/shell');
jest.mock('../utils/writeSpecs');

describe('BUMP command', () => {
  it("leaves package.json untouched if it doesn't depend on dep", async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await bump(['mady@18'], {
      src: 'test/fixtures/packages/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(0);
  });

  it('modifies package.json if it depends on dep', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await bump(['ext-one@18'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();
  });

  it('modifies a dep version everywhere it may appear', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await bump(['ext-two@18'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(1);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
  });

  it('modifies multiple deps', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await bump(['ext-one@18', 'ext-two@18'], {
      src: 'test/fixtures/packagesCustomLinks/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();
  });

  it('uses current version for internal deps', async () => {
    const writeSpecs = require('../utils/writeSpecs').default;
    await bump(['oao-c'], {
      src: 'test/fixtures/packages/*',
    });
    expect(writeSpecs.mock.calls).toHaveLength(2);
    expect(writeSpecs.mock.calls[0][1]).toMatchSnapshot();
    expect(writeSpecs.mock.calls[1][1]).toMatchSnapshot();
  });
});
