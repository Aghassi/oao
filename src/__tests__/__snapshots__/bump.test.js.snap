// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BUMP command modifies a dep version everywhere it may appear 1`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packagesCustomLinks/oao-c/package.json",
  Object {
    "dependencies": Object {
      "ext-one": "*",
      "oao": "*",
      "timm": "1.x",
    },
    "devDependencies": Object {
      "ext-two": "18",
      "oao-b": "*",
      "xxl": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-c",
    "peerDependencies": Object {
      "ext-two": "18",
      "oao-b": "*",
    },
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command modifies multiple deps 1`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packagesCustomLinks/oao-b/package.json",
  Object {
    "dependencies": Object {
      "ext-one": "18",
      "oao": "*",
      "timm": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-b",
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command modifies multiple deps 2`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packagesCustomLinks/oao-c/package.json",
  Object {
    "dependencies": Object {
      "ext-one": "18",
      "oao": "*",
      "timm": "1.x",
    },
    "devDependencies": Object {
      "ext-two": "18",
      "oao-b": "*",
      "xxl": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-c",
    "peerDependencies": Object {
      "ext-two": "18",
      "oao-b": "*",
    },
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command modifies package.json if it depends on dep 1`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packagesCustomLinks/oao-b/package.json",
  Object {
    "dependencies": Object {
      "ext-one": "18",
      "oao": "*",
      "timm": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-b",
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command modifies package.json if it depends on dep 2`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packagesCustomLinks/oao-c/package.json",
  Object {
    "dependencies": Object {
      "ext-one": "18",
      "oao": "*",
      "timm": "1.x",
    },
    "devDependencies": Object {
      "ext-two": "*",
      "oao-b": "*",
      "xxl": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-c",
    "peerDependencies": Object {
      "ext-two": "*",
      "oao-b": "*",
    },
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command uses current version for internal deps 1`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packages/oao-d/package.json",
  Object {
    "dependencies": Object {
      "oao": "*",
      "timm": "1.x",
    },
    "devDependencies": Object {
      "oao-b": "*",
      "oao-c": "^0.1.0",
      "xxl": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-d",
    "peerDependencies": Object {},
    "scripts": Object {},
    "version": "0.1.0",
  },
]
`;

exports[`BUMP command uses current version for internal deps 2`] = `
Array [
  "/Users/guillermo/Prog/guigrpa/oao/test/fixtures/packages/oao-priv/package.json",
  Object {
    "dependencies": Object {
      "oao": "*",
      "timm": "1.x",
    },
    "devDependencies": Object {
      "oao-b": "*",
      "oao-c": "^0.1.0",
      "xxl": "1.x",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "oao-priv",
    "peerDependencies": Object {},
    "private": true,
    "version": "0.1.0",
  },
]
`;
